Capstone Project

In [1]:
!pip install beautifulsoup4

import urllib
from bs4 import BeautifulSoup
import pandas as pd

html = urllib.request.urlopen("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M").read()
soup = BeautifulSoup(html, 'html.parser')


table = soup.find('table', {'class': 'wikitable sortable'})

columns = [th.text.replace('\n', '') for th in table.find('tr').find_all('th')]
columns

['Postcode', 'Borough', 'Neighbourhood']

In [17]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


In [2]:

trs = table.find_all('tr')[1:]
rows = list()
for tr in trs:
    rows.append([td.text.replace('\n', '').replace('\xa0', '') for td in tr.find_all('td')])
rows[:5]

df = pd.DataFrame(data=rows, columns=columns)
df.head()

Postcode           Borough     Neighbourhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

In [3]:
df = df.loc[df['Borough']!='Not assigned']

df = df.groupby('Postcode').agg({'Borough':'first', 
                             'Neighbourhood': ', '.join}).reset_index()


df.loc[df['Neighbourhood']=='Not assigned']

df.loc[df['Neighbourhood']=='Not assigned','Neighbourhood']=df['Borough']

In [4]:
df

Postcode      Borough                                      Neighbourhood
0        M1B  Scarborough                                     Rouge, Malvern
1        M1C  Scarborough             Highland Creek, Rouge Hill, Port Union
2        M1E  Scarborough                  Guildwood, Morningside, West Hill
3        M1G  Scarborough                                             Woburn
4        M1H  Scarborough                                          Cedarbrae
..       ...          ...                                                ...
98       M9N         York                                             Weston
99       M9P    Etobicoke                                          Westmount
100      M9R    Etobicoke  Kingsview Village, Martin Grove Gardens, Richv...
101      M9V    Etobicoke  Albion Gardens, Beaumond Heights, Humbergate, ...
102      M9W    Etobicoke                                          Northwest

[103 rows x 3 columns]

In [5]:
file_name='http://cocl.us/Geospatial_data'
df2=pd.read_csv(file_name)

df = pd.merge(df,df2, how='left', left_on=['Postcode'], right_on=['Postal Code'], validate="one_to_one")

df.drop('Postal Code',axis = 1, inplace = True)

In [8]:
df

Postcode      Borough                                      Neighbourhood  \
0        M1B  Scarborough                                     Rouge, Malvern   
1        M1C  Scarborough             Highland Creek, Rouge Hill, Port Union   
2        M1E  Scarborough                  Guildwood, Morningside, West Hill   
3        M1G  Scarborough                                             Woburn   
4        M1H  Scarborough                                          Cedarbrae   
..       ...          ...                                                ...   
98       M9N         York                                             Weston   
99       M9P    Etobicoke                                          Westmount   
100      M9R    Etobicoke  Kingsview Village, Martin Grove Gardens, Richv...   
101      M9V    Etobicoke  Albion Gardens, Beaumond Heights, Humbergate, ...   
102      M9W    Etobicoke                                          Northwest   

      Latitude  Longitude  
0    43.806686 -79.194353  
1    43.784535 -79.160497  
2    43.763573 -79.188711  
3    43.770992 -79.216917  
4    43.773136 -79.239476  
..         ...        ...  
98   43.706876 -79.518188  
99   43.696319 -79.532242  
100  43.688905 -79.554724  
101  43.739416 -79.588437  
102  43.706748 -79.594054  

[103 rows x 5 columns]

In [20]:
CLIENT_ID = 'JEMAFN1CFNHADJ4FPYSLQYNZZ3NPOD3ZCQ35HUOZSGJ1MNDJ' # your Foursquare ID
CLIENT_SECRET = 'UNL3TROJTQRKYZBOBIFDU4SCF0W0TGG0QYZDPPRMFQ2CPCCT' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: JEMAFN1CFNHADJ4FPYSLQYNZZ3NPOD3ZCQ35HUOZSGJ1MNDJ
CLIENT_SECRET:UNL3TROJTQRKYZBOBIFDU4SCF0W0TGG0QYZDPPRMFQ2CPCCT


In [19]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(df['Borough'].unique()),
        df.shape[0]
    )
)

The dataframe has 11 boroughs and 103 neighborhoods.


In [11]:
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

In [18]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[43.741667, -79.373333], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df['Latitude'], df['Longitude'], df['Borough'], df['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [26]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [27]:
# type your answer here
LIMIT = 100 
toronto_venues = getNearbyVenues(names=df['Neighbourhood'],
                                   latitudes=df['Latitude'],
                                   longitudes=df['Longitude']
                                  )



Rouge, Malvern
Highland Creek, Rouge Hill, Port Union
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
East Birchmount Park, Ionview, Kennedy Park
Clairlea, Golden Mile, Oakridge
Cliffcrest, Cliffside, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Scarborough Town Centre, Wexford Heights
Maryvale, Wexford
Agincourt
Clarks Corners, Sullivan, Tam O'Shanter
Agincourt North, L'Amoreaux East, Milliken, Steeles East
L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
Silver Hills, York Mills
Newtonbrook, Willowdale
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Flemingdon Park, Don Mills South
Bathurst Manor, Downsview North, Wilson Heights
Northwood Park, York University
CFB Toronto, Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Woodbine Gardens, Parkview Hill
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto
The Danforth West, 

In [17]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


In [18]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5e0038960be7b4001bef1334'},
 'response': {'warning': {'text': "There aren't a lot of results near you. Try something more general, reset your filters, or expand the search area."},
  'headerLocation': 'Malvern',
  'headerFullLocation': 'Malvern, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 1,
  'suggestedBounds': {'ne': {'lat': 43.8111863045, 'lng': -79.18812958073042},
   'sw': {'lat': 43.80218629549999, 'lng': -79.2005772192696}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4bb6b9446edc76b0d771311c',
       'name': "Wendy's",
       'location': {'crossStreet': 'Morningside & Sheppard',
        'lat': 43.80744841934756,
        'lng': -79.19905558052072,
        'labeledLatLngs': [{'label': 'display'

In [28]:
print(toronto_venues.shape)
toronto_venues.head()

(2210, 7)


Neighborhood  Neighborhood Latitude  \
0                          Rouge, Malvern              43.806686   
1  Highland Creek, Rouge Hill, Port Union              43.784535   
2  Highland Creek, Rouge Hill, Port Union              43.784535   
3       Guildwood, Morningside, West Hill              43.763573   
4       Guildwood, Morningside, West Hill              43.763573   

   Neighborhood Longitude                            Venue  Venue Latitude  \
0              -79.194353                          Wendy's       43.807448   
1              -79.160497                RIGHT WAY TO GOLF       43.785177   
2              -79.160497            Royal Canadian Legion       43.782533   
3              -79.188711  Swiss Chalet Rotisserie & Grill       43.767697   
4              -79.188711                G & G Electronics       43.765309   

   Venue Longitude        Venue Category  
0       -79.199056  Fast Food Restaurant  
1       -79.161108           Golf Course  
2       -79.163085                   Bar  
3       -79.189914           Pizza Place  
4       -79.191537     Electronics Store

In [36]:
set(df['Borough'])

{'Central Toronto',
 'Downtown Toronto',
 'East Toronto',
 'East York',
 'Etobicoke',
 'Mississauga',
 'North York',
 "Queen's Park",
 'Scarborough',
 'West Toronto',
 'York'}

In [37]:
# type your answer here
LIMIT = 500 # Maximum is 100
cities = ['Central Toronto', ##
 'Downtown Toronto',
 'East Toronto',##
 'East York',
 'Etobicoke',
 'Mississauga',
 'North York',
 "Queen's Park",
 'Scarborough',
 'West Toronto',
 'York']
results = {}
for city in cities:
    url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&near={}&limit={}&categoryId={}'.format(
        CLIENT_ID, 
        CLIENT_SECRET, 
        VERSION, 
        city,
        LIMIT,
        "4bf58dd8d48988d16d941735") # CAFE CATEGORY ID
    results[city] = requests.get(url).json()

WE FOUND THAT THERE IS NO RESPONSE IN CENTRAL TORONTO AND EAST TORONTO, SO WE REMOVE THEM FROM THE LIST

In [54]:
cities = [
 'Downtown Toronto',
 'East York',
 'Etobicoke',
 'Mississauga',
 'North York',
 "Queen's Park",
 'Scarborough',
 'West Toronto',
 'York']

In [55]:
df_venues={}
for city in cities:
    venues = json_normalize(results[city]['response']['groups'][0]['items'])
    df_venues[city] = venues[['venue.name', 'venue.location.address', 'venue.location.lat', 'venue.location.lng']]
    df_venues[city].columns = ['Name', 'Address', 'Lat', 'Lng']

Check out the cafe amount in each city

In [85]:
maps = {}


for city in cities:
    city_lat = results[city]['response']['geocode']['center']['lat']

           
    city_lng = results[city]['response']['geocode']['center']['lng']
                 
    maps[city] = folium.Map(location=[city_lat, city_lng], zoom_start=13)

    # add markers to map
    for lat, lng, label in zip(df_venues[city]['Lat'], df_venues[city]['Lng'], df_venues[city]['Name']):
        label = folium.Popup(label, parse_html=True)
        folium.CircleMarker(
            [lat, lng],
            radius=5,
            popup=label,
            color='blue',
            fill=True,
            fill_color='#3186cc',
            fill_opacity=0.7,
            parse_html=False).add_to(maps[city])  

    print(f"Total number of CAFE in {city} = ", results[city]['response']['totalResults'])
    print("Showing Top 100")

Total number of CAFE in Downtown Toronto =  111
Showing Top 100
Total number of CAFE in East York =  67
Showing Top 100
Total number of CAFE in Etobicoke =  93
Showing Top 100
Total number of CAFE in Mississauga =  87
Showing Top 100
Total number of CAFE in North York =  103
Showing Top 100
Total number of CAFE in Queen's Park =  12
Showing Top 100
Total number of CAFE in Scarborough =  58
Showing Top 100
Total number of CAFE in West Toronto =  82
Showing Top 100
Total number of CAFE in York =  47
Showing Top 100


In [86]:
maps[cities[0]]

In [87]:
maps[cities[1]]

In [88]:
maps[cities[2]]

In [89]:
maps[cities[3]]

In [90]:
maps[cities[4]]

In [91]:
maps[cities[5]]

In [92]:
maps[cities[6]]

In [93]:
maps[cities[7]]

In [94]:
maps[cities[8]]

We can see that downtown Toronto and North York has the most cafe in town. Let's have a concrete measure of this density.For this I will use some basic statistics. I will get the mean location of the cafe which should be near to most of them if they are really dense or far if not.And I will take the average of the distance of the venues to the mean coordinate

In [96]:
maps = {}
for city in cities:
    city_lat = results[city]['response']['geocode']['center']['lat']

           
    city_lng = results[city]['response']['geocode']['center']['lng']
                 
    maps[city] = folium.Map(location=[city_lat, city_lng], zoom_start=13)
    
    venues_mean_coor = [df_venues[city]['Lat'].mean(), df_venues[city]['Lng'].mean()] 
    # add markers to map
    for lat, lng, label in zip(df_venues[city]['Lat'], df_venues[city]['Lng'], df_venues[city]['Name']):
        label = folium.Popup(label, parse_html=True)
        folium.CircleMarker(
            [lat, lng],
            radius=5,
            popup=label,
            color='blue',
            fill=True,
            fill_color='#3186cc',
            fill_opacity=0.7,
            parse_html=False).add_to(maps[city])
        folium.PolyLine([venues_mean_coor, [lat, lng]], color="green", weight=1.5, opacity=0.5).add_to(maps[city])
    
    label = folium.Popup("Mean Co-ordinate", parse_html=True)
    folium.CircleMarker(
        venues_mean_coor,
        radius=10,
        popup=label,
        color='green',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(maps[city])

    print(city)
    print("Mean Distance from Mean coordinates")
    print(np.mean(np.apply_along_axis(lambda x: np.linalg.norm(x - venues_mean_coor),1,df_venues[city][['Lat','Lng']].values)))

Downtown Toronto
Mean Distance from Mean coordinates
0.004928355936759374
East York
Mean Distance from Mean coordinates
0.014176758888982533
Etobicoke
Mean Distance from Mean coordinates
0.0613207492389694
Mississauga
Mean Distance from Mean coordinates
0.06393634263776303
North York
Mean Distance from Mean coordinates
0.05052552660236895
Queen's Park
Mean Distance from Mean coordinates
0.01838488690565628
Scarborough
Mean Distance from Mean coordinates
0.05717454980627016
West Toronto
Mean Distance from Mean coordinates
0.015961898750024964
York
Mean Distance from Mean coordinates
0.03082937873623301


In [105]:
maps[cities[8]] #York

We now see that Downtown Toronto is his best option. Our tourist's best interest would be to book a hotel near that mean coordinate to surround himself with the 111 cafe there!!